### PR TITLE
Handle role members for trust domain

### PR DESCRIPTION
**Pull request by dxdeidara**
This pr adds the feature for fetching role members for the role that has delegated domain. It has the same behavior of how zms fetches the role members. This pr also adds the test case for parsing trustdomain and authorizing user that is a member of corresponding role in trustdomain.

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by mcieplak** [comment]
cc @prabushyam

File: zpeclient.go [hunk 1]
```diff
@@ -163,9 +173,69 @@ func (c *Cache) parseData(item *v1.AthenzDomain) (roleMappings, error) {
...
+				c.log.Println("policy in Contents.Policies is nil")
+				continue
+			}
+			for _, assertion := range policy.Assertions {
+				// check if policy contains action "assume_role", and resource matches with delegated role name
+				if assertion.Action == "assume_role" {
+					// form correct role name
+					matched, err := regexp.MatchString("^"+roleReplacer.Replace(assertion.Resource)+"$", roleName)
+					if err != nil {
+						c.log.Println("string matching failed with err: ", err)
```
Comment (mcieplak) [inline zpeclient.go]: We should continue here.

File: zpeclient_test.go [hunk 2]
```diff
@@ -9,13 +9,37 @@ import (
...
-	domainName = "home.domain"
-	username   = "user.name"
+	domainName      = "home.domain"
+	username        = "user.name"
+	trustDomainName = "test.delegated.domain"
+	trustusername   = "trustuser.name"
+)
+
+var (
+	ad = &v1.AthenzDomain{
```
Comment (mcieplak) [inline zpeclient_test.go]: Should we use the `getFakeAthenzDomains` and `getFakeTrustAthenzDomains` functions to set ad / ad1?

File: zpeclient_test.go [hunk 3]
```diff
@@ -76,7 +112,72 @@ func getFakeDomain() zms.SignedDomain {
...
+			},
+			Services: []*zms.ServiceIdentity{},
+			Entities: []*zms.Entity{},
+		},
+		KeyId:     "colo-env-1.1",
+		Signature: "signature",
+	}
+}
+
+func getFakeTrustAthenzDomains() *v1.AthenzDomain {
```
Comment (mcieplak) [inline zpeclient_test.go]: Could we remove the `s` at the end in the function name and also do the same for the `getFakeAthenzDomains` function as well.

File: zpeclient.go [hunk 4]
```diff
@@ -163,9 +173,69 @@ func (c *Cache) parseData(item *v1.AthenzDomain) (roleMappings, error) {
...
 }
 
+// processTrustDomain -  in delegated domain check assume_role action in policy that contains current role as a resource, return role's member list
+func (c *Cache) processTrustDomain(trust zms.DomainName, item *v1.AthenzDomain, roleName string) []*zms.RoleMember {
+	var res []*zms.RoleMember
+	// handle case which crIndexInformer is not initialized at the beginning, return directly.
+	if c.crIndexInformer == nil {
+		return res
+	}
+	if trust != "" {
```
[the remark this answers is not in the document]
Comment (dxdeidara) [reply zpeclient.go]: This is actually a left over when I was implementing recursion in this func. I will remove it.

File: go.mod [hunk 5]
```diff
@@ -4,11 +4,11 @@ go 1.12
 
 require (
 	github.com/ardielle/ardielle-go v1.5.2
+	github.com/davecgh/go-spew v1.1.1
```
[the remark this answers is not in the document]
Comment (dxdeidara) [reply go.mod]: done

File: zpeclient.go [hunk 6]
```diff
@@ -99,11 +102,18 @@ func (c *Cache) parseData(item *v1.AthenzDomain) (roleMappings, error) {
 			continue
 		}
 		roleName := string(role.Name)
+		// Handle trust domains: if string(role.Trust) != ""
+		// For a trusted domain, role object will only contain following items:
+		// modified, name, trust
+		if role.Trust != "" {
+			role.RoleMembers = c.processTrustDomain(role.Trust, item, roleName)
```
[the remark this answers is not in the document]
Comment (dxdeidara) [reply zpeclient.go]: I left a comment in the code

File: zpeclient_test.go (hunk 2 above)
Comment (dxdeidara) [reply zpeclient_test.go]: I moved the code piece to NewCache() method to avoid code duplication

File: zpeclient_test.go (hunk 3 above)
Comment (dxdeidara) [reply zpeclient_test.go]: done

File: zpeclient_test.go [hunk 7]
```diff
@@ -345,6 +465,15 @@ func TestAuthorize(t *testing.T) {
 		t.Error("Wrong authorization result, authorization should pass.")
 	}
 
+	// grant trust user access
+	res, err = privateCache.authorize(trustusername, check)
```
[the remark this answers is not in the document]
Comment (dxdeidara) [reply zpeclient_test.go]: added

File: zpeclient_test.go [hunk 8]
```diff
@@ -9,16 +9,40 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
 )
 
 const (
-	domainName = "home.domain"
-	username   = "user.name"
+	domainName      = "home.domain"
+	username        = "user.name"
+	trustDomainName = "test.delegated.domain"
```
[the remark this answers is not in the document]
Comment (dxdeidara) [reply zpeclient_test.go]: changed

File: zpeclient_test.go [hunk 9]
```diff
@@ -58,6 +82,18 @@ func getFakeDomain() zms.SignedDomain {
 							Modified: &timestamp,
 							Name:     zms.ResourceName(domainName + ":policy.admin"),
 						},
+						{
+							Assertions: []*zms.Assertion{
+								{
+									Role:     domainName + ":role.delegated",
+									Resource: domainName + ":*",
```
[the remark this answers is not in the document]
Comment (dxdeidara) [reply zpeclient_test.go]: Switched to the specific one from *

File: go.mod [hunk 10]
```diff
@@ -6,9 +6,8 @@ require (
 	github.com/ardielle/ardielle-go v1.5.2
 	github.com/stretchr/testify v1.3.0
 	github.com/yahoo/athenz v1.8.24
-	github.com/yahoo/k8s-athenz-istio-auth v0.0.0-20190717030003-19444f7f59a2
-	k8s.io/api v0.0.0-20190222213804-5cb15d344471
-	k8s.io/apimachinery v0.0.0-20190629003722-e20a3a656cff
+	github.com/yahoo/k8s-athenz-syncer v0.1.1
+	k8s.io/api v0.0.0-20190817221950-ebce17126a01
+	k8s.io/apimachinery v0.0.0-20190817221809-bf4de9df677c
 	k8s.io/client-go v10.0.0+incompatible
```
Comment (prabushyam) [inline go.mod]: can we update client-go to v12 (compat with 1.15)?

File: zpeclient.go [hunk 11]
```diff
@@ -99,11 +102,23 @@ func (c *Cache) parseData(item *v1.AthenzDomain) (roleMappings, error) {
...
 		roleName := string(role.Name)
+		// Handle trust domains: if string(role.Trust) != ""
+		// For a trusted domain, role object will only contain following items:
+		// modified, name, trust.
+		// Note: logic may change depend on how signature validation is implemented.
+		if role.Trust != "" {
+			var err error
+			role.RoleMembers, err = c.processTrustDomain(role.Trust, item, roleName)
+			if err != nil {
+				c.log.Printf("Error occurred when processing trust domain. Error: %v", err)
```
Comment (prabushyam) [inline zpeclient.go]: can we add a continue statement here ? (since we shouldn't further process this role for any members)

Comment (dxdeidara) [reply zpeclient.go]: done

File: go.mod (hunk 10 above)
Comment (dxdeidara) [reply go.mod]: updated

**Pull request merged**